Proxifier should use superclass when proxifying a weld proxy

If the proxyfier tries to proxify a class that is already proxified, javassist throws a exception. Something like:

```
Caused by: java.lang.RuntimeException: duplicate method: getHandler in br.com.caelum.brutal.infra.rss.converter.OndeTrabalharRSSJob$Proxy$_$$_WeldClientProxy_$$_javassist_20
    at javassist.util.proxy.ProxyFactory.createClass3(ProxyFactory.java:509)
    at javassist.util.proxy.ProxyFactory.createClass2(ProxyFactory.java:486)
    at javassist.util.proxy.ProxyFactory.createClass1(ProxyFactory.java:422)
    at javassist.util.proxy.ProxyFactory.createClass(ProxyFactory.java:394)
    at br.com.caelum.vraptor4.proxy.JavassistProxifier.proxify(JavassistProxifier.java:84)
```

So, before creating the proxy, the proxyfier should verify if the instance is already a proxy and get its superclass.

--- a/vraptor-core/src/main/java/br/com/caelum/vraptor4/proxy/JavassistProxifier.java
+++ b/vraptor-core/src/main/java/br/com/caelum/vraptor4/proxy/JavassistProxifier.java
@@ -71,14 +71,19 @@ public class JavassistProxifier
     }
 
     @Override
+    @SuppressWarnings("rawtypes")
 	public <T> T proxify(Class<T> type, MethodInvocation<? super T> handler) {
         final ProxyFactory factory = new ProxyFactory();
         factory.setFilter(IGNORE_BRIDGE_AND_OBJECT_METHODS);
+		Class rawType = type;
+        if (isProxyClass(type)) {
+        	rawType = type.getSuperclass();
+        }
 
         if (type.isInterface()) {
-            factory.setInterfaces(new Class[] { type });
+            factory.setInterfaces(new Class[] { rawType });
         } else {
-            factory.setSuperclass(type);
+            factory.setSuperclass(rawType);
         }
 
         Class<?> proxyClass = factory.createClass();
@@ -93,8 +98,12 @@ public class JavassistProxifier
 
     @Override
 	public boolean isProxy(Object o) {
-        return o != null && ProxyObject.class.isAssignableFrom(o.getClass());
+        return o != null && isProxyClass(o.getClass());
     }
+
+	private boolean isProxyClass(Class<? extends Object> clazz) {
+		return ProxyObject.class.isAssignableFrom(clazz);
+	}
     
     private <T> void setHandler(Object proxyInstance, final MethodInvocation<? super T> handler) {
         ProxyObject proxyObject = (ProxyObject) proxyInstance;

--- a/vraptor-core/src/test/java/br/com/caelum/vraptor4/proxy/JavassistProxifierTest.java
+++ b/vraptor-core/src/test/java/br/com/caelum/vraptor4/proxy/JavassistProxifierTest.java
@@ -20,6 +20,7 @@ package br.com.caelum.vraptor4.proxy;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.fail;
 
 import java.lang.reflect.Method;
@@ -29,19 +30,12 @@ import net.vidageek.mirror.dsl.Mirror;
 import org.junit.Before;
 import org.junit.Test;
 
-import br.com.caelum.vraptor4.proxy.JavassistProxifier;
-import br.com.caelum.vraptor4.proxy.MethodInvocation;
-import br.com.caelum.vraptor4.proxy.ObjenesisInstanceCreator;
-import br.com.caelum.vraptor4.proxy.Proxifier;
-import br.com.caelum.vraptor4.proxy.ProxyInvocationException;
-import br.com.caelum.vraptor4.proxy.SuperMethod;
-
 /**
  * @author Fabio Kung
  */
 public class JavassistProxifierTest {
 
-    private Proxifier proxifier;
+    private JavassistProxifier proxifier;
     
     @Before
     public void setUp() throws Exception {
@@ -149,6 +143,20 @@ public class JavassistProxifierTest {
 			}
 		}
     }
+    
+    @Test
+	public void shouldConsiderSuperclassWhenProxifiyngProxy() throws Exception {
+		MethodInvocation<C> handler = new MethodInvocation<C>() {
+			@Override
+			public Object intercept(C proxy, Method method, Object[] args,
+					SuperMethod superMethod) {
+				return null;
+			}
+		};
+		C firstProxy = proxifier.proxify(C.class, handler);
+		C secondProxy = proxifier.proxify(firstProxy.getClass(), handler);
+		assertEquals(firstProxy.getClass(), secondProxy.getClass());
+	}
 
 	public static class A<T> {
 		public T getT(T t) { return t; }

